Fix GH-19685: Reject out-of-range blocks/work values in bzip2.compress filter

alexandre-daubois opened the pull request:
Fix #19685 

When `blocks` or `work` have invalid values, it currently let the function execution continue. Instead, we should free some resources and return null to avoid the segfault.

alexandre-daubois committed:
Fix GH-19685: Reject out-of-range blocks/work values in bzip2.compress filter

alexandre-daubois commented on NEWS:
```diff
@@ -2,6 +2,10 @@ PHP                                                                        NEWS
 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
 ?? ??? ????, PHP 8.3.26
 
+- Bz2:
```
Ah yes, thanks!

devnexen commented on ext/bz2/bz2_filter.c:
```diff
@@ -367,6 +367,10 @@ static php_stream_filter *php_bz2_filter_create(const char *filtername, zval *fi
 					zend_long blocks = zval_get_long(tmpzval);
 					if (blocks < 1 || blocks > 9) {
 						php_error_docref(NULL, E_WARNING, "Invalid parameter given for number of blocks to allocate (" ZEND_LONG_FMT ")", blocks);
+						pefree(data->strm.next_in, persistent);
```
nit: only if you want, on master only maybe, making a goto label to cover the 3 cases.

bukka commented:
It should be ok in this case but breaking filter chain is sometimes not safe (might cause breaks) so I will need to review it properly - I should be properly looking to in 2 or 3 weeks time.

bukka reviewed (changes requested):
So I just spent some time to getting to know bzip2 extension to see if I'm missing anything by any chance but I'm not sure I understand what the point of this PR is.

You claim (in the PR title) that this is fixes the segfault but from the stacktrace in https://github.com/php/php-src/issues/19685#issuecomment-3250643861, this looks to me that the fix for that is in https://github.com/php/php-src/pull/19708 because it's segfaulting due to trying to use ops write on data stream (that has got write ops NULL). That bug GH-19685 is really just a duplicate of GH-19705 - the only difference is that it's using bzip2 filter which also as got some buffered rubbish to write. Basically if you changed the `blocks` to 4 and `work` to 0, it would still segfault so this really doesn't fix anything and it's completely unrelated to that bug. In fact this is not a bug at all but just a BC break because currently it's just warning and then using defaults. But you changed it to not use defaults and fails. This should really go through deprecation.

So at least you need to change the branch, titles and description and then add it to the deprecation RFC for 8.6.

iluuu1994 commented on NEWS:
```diff
@@ -20,6 +20,10 @@ PHP                                                                        NEWS
   . Added NUL-byte validation to BCMath functions. (jorgsowa)
 
+- Bz2:
+  . Fixed bug GH-19685 (Segfault when bzip2 filter has invalid parameters).
+    (alexandre-daubois)
```
I think that's the format we use even for features. At least I've never seen a different format when referring to GitHub issues.

alexandre-daubois replied:
Indeed, but I can remove "bug" to indicate this fixes the Github issue itself if necessary

iluuu1994 replied:
I think the format should remain the same. The format is even documented here: https://github.com/php/php-src/blob/master/CONTRIBUTING.md#news

alexandre-daubois commented:
Is it good to merge? 🙂

bukka commented:
I will need to find time to check it out. Last time it didn't make much sense and wasn't fixing any segfault. So looks like at least title and the commit should change.

alexandre-daubois commented:
Indeed, I updated the commit message + title. Also, I forgot about https://github.com/php/php-src/pull/19702/changes#r2322984175, which is now addressed

bukka commented:
The change looks good for master but is this really a fix for #19685 ? Wasn't it already fixed by https://github.com/php/php-src/pull/19708 ?

bukka commented:
In other words please check if you can recreate it after that fix - it should no longer crash. If that's the case, then, please, modify the title and commit message again and then feel free to merge it...
